### ISSUE TITLE
Can you do multiple builds of the same repo concurrently on a single host machine?

### ISSUE THREAD

**Issue by zjturner** [facebook/buck2#449]
I tried this with a simple project and it "appeared to work", but I am unsure if this should be relied on.  I can reason that many parts of buck2 are necessarily re-entrant by design since it's remote first, and a remote server has to be able to service many requests concurrently.  But from a single machine, I'm less sure if the buck2 daemon and file system code is guaranteed to be concurrency-aware.  For example, what if I do `buck2 build //foo:bar` and `buck2 build //foo:baz` concurrently on a single machine?  Or, a different possibility, running `buck2 build --target-platform=foo` and `buck2 build --target-platform=bar` concurrently.

Is this supported?

**Comment by cjhopman** [comment]
Multiple concurrent builds should be mostly okay. Currently, errors encountered might print some of their information only to one of the builds (both would still get the error if appropriate) but that's actively being fixed.

concurrent builds with different --target-platform should also be okay.

concurrent builds with different `-c` flags or with different file state or a handful of other things aren't okay, but that case should be detected and blocked. actually, maybe that blocking requires a particular buckconfig, it should be default but idk if it is @wendy728 would probably know

**Comment by zjturner** [comment]
How are races for shared work handled when doing this?  For example, suppose I have a V shaped dependency where A and B both depend on C.  I run `buck2 build A` and then concurrently run `buck2 build B`.  These are going to race for triggering the build of C.  Is this just a natural fallout of the design of going through a single instance of buck2d?  So say A comes in first and C starts building and A naturally waits for C, and then B's request comes through and buck2 just has B wait for C no differently than A which is already waiting for C?

**Comment by wendy728** [comment]
> maybe that blocking requires a particular buckconfig

We don't need to specify a specific buckconfig anymore (cleaned that up a while back). 

@zjturner In this V shaped scenario, C would become a "shared task" for both A and B, and should be evaluated once. 

See [here](https://github.com/facebook/buck2/blob/main/docs/users/faq/common_issues.md#are-multiple-concurrent-commands-supported) for more info about concurrent commands.

**Comment by thoughtpolice** [comment]
Small note, but while the title talks about "same machine", I suspect you *probably* mean the same *repository* or build directory, right? But just to round out the issue, I think that as long as the buck2 isolation directories are different &mdash; and if you're running in two separate directories, they are by definition different &mdash; there is no need to worry about intra-machine builds interfering at all. So, if you're running two builds from, say, two different checkouts of a codebase in your CI system, you should be fine.

I will note that I have been running `buck2 build` concurrently with and without different mode/config files for a while, and in that case as Wendy noted you'll see something like this when buck2 blocks on what are effectively `buckconfig` changes where things would otherwise blow up I guess:

```
austin@GANON$ just build-all @mode//rel
buck2 build @mode//rel root//... third-party//... cellar//... tilde//... 
Build ID: 8050fb78-5a3d-4cd6-ab7a-8723d6e76e7b
Network: Up:   0B             Down:  82MiB 6.9MiB/s  (GRPC-SESSION-ID)
Command: build.                                                                                     Remaining: 1/2. Time elapsed: 7.3s
--------------------------------------------------------------------------------------------------------------------------------------
Synchronizing buck2 internal state [Waiting for command [buck2 build @mode//dev root//... third-party//... cellar//... tilde//... 7.3s
```

So I suspect all the cases you're worried about are fine.

**Comment by zjturner** [comment]
Yes, I meant same directory.  Good to know that this works, thanks.  From @cjhopman 's original response (that case should be detected and blocked) I interpreted blocked as "prevented", like buck2 should give you an error if you do that.  But now after Wendy's response and reading the linked page, I understand it to mean that it will still work just fine, but one invocation will wait for the other invocation to finish.

Basically, it sounds like there's no case where this will simply result in a broken / incorrect build.  That was my main concern.
